### ISSUE TITLE
Upgrading my project to version 2.4.0+ multiplies x10 output time

### ISSUE THREAD

**Issue by begonaalvarezd** [sphinx-doc/sphinx#7223]
I would like to upgrade my project to the latest Sphinx version but I have encountered myself that if I upgrade to version 2.4.0+, when the project is in phase "writing output" it takes >10min to finish, when any other version (eg 2.3.1 and lower until 1.3.1) takes around 100s.

I have discovered that the bottleneck is in the sidebar template: 
```
<div id="toctree">
{{ toctree(maxdepth=6|toint, collapse=False,includehidden=True|tobool) }}
</div>
```
If I dont add a sidebar then everything is smooth as always, but I do need the sidebar with the entire tree for navigation purporses.

Touching here and there the installed package I have discovered that removing in file `environment/adapters/toctree.py` line 218 `entries.extend(children)` stops this slow performance, but of course the toctree creation disappears (in version 2.4.0).

My project is quite big, it has over 400 source files so I understand building the toctree is time consuming, but honestly upgrading and multiplying by 10 the output time...stops me from upgrading to the latest version. 

My source files didnt changed from one version to another and I am using  Python 3.7.6. 

I cannot share with you the source files because the project is private but in case it helps you here goes a link to the current documentation site https://doc.synapticon.com/ (which uses Python 2.7.15 and Sphinx 1.3.1, hence the urge to upgrade) and the conf file of the current under progress upgraded project (to new version):

```
import sys, os, re
import sphinx_bootstrap_theme

project = 'sncn_documentation'
copyright = u''
author = 'Begoña Álvarez - Boxfish Studio SL'
sys.path.append("./ext")
extensions = [
    'sphinx.ext.ifconfig',
    'sphinx.ext.mathjax',
    'div',
    'tabselection',
]
templates_path = ['_templates']
exclude_patterns = [
    '_build', '_static', 'ext', '_templates', 'Thumbs.db', 
    '.DS_Store', '.support*','.sources*','.doxygen*', 
    'README.rst','**/README.rst','**/pseudocode.rst',
    '**/CHANGELOG.rst', '*.#*'
    ]
html_theme = 'bootstrap'
html_static_path = ['_static']
html_title = "Synapticon Documentation"

html_sidebars = {'**/**': ['toctree.html']}
html_show_sourcelink = False
```

**Comment by jfbu** [comment]
Thanks for reporting. I edited the title to refer to `2.4.0` rather.

**Comment by tk0miya** [comment]
Thank you for reporting and sorry for inconvenience. I found the reason of the slow. I'll merge #7224 and release bug fix version later.

**Comment by tk0miya** [comment]
Fixed by #7224 
Thank you for reporting!
